### ISSUE TITLE
FP8 Support for FlexAttention

### ISSUE THREAD

**Issue by harveyp123** [pytorch/pytorch#151695]
### 🚀 The feature, motivation and pitch

For FlexAttention, is it possible to add FP8 support for inference? FP8+Flash Attention 3 seems to be a great success, it will be great if FlexAttention can support FP8 inference

**Comment by drisspg** [comment]
The short answer is that this works today:
```py
import torch
from torch.nn.attention.flex_attention import flex_attention
from functools import partial
from typing import Callable
import logging
from transformer_nuggets.utils.benchmark import benchmark_cuda_function_in_microseconds


logging.getLogger("transformer_nuggets").setLevel(logging.INFO)
Tensor = torch.Tensor


def per_tensor_scaling(score, b, h, q_idx, k_idx, q_scale, k_scale):
    return score.to(torch.float32) * q_scale * k_scale


def scaled_flex(
    query: Tensor,
    key: Tensor,
    value: Tensor,
    q_scale: Tensor,
    k_scale: Tensor,
    v_scale: Tensor,
    scale_func: Callable,
    **kwargs
):
    scale_func_partial = partial(scale_func, q_scale=q_scale, k_scale=k_scale)
    out = flex_attention(query, key, value, score_mod=scale_func_partial, **kwargs)
    out = out.to(torch.bfloat16)
    out *= v_scale
    return out


compiled_scale_flex = torch.compile(scaled_flex, fullgraph=True, dynamic=False, mode="max-autotune-no-cudagraphs")
compiled_flex = torch.compile(flex_attention, fullgraph=True, dynamic=False, mode="max-autotune-no-cudagraphs")

torch._dynamo.config.cache_size_limit = 1000

make_tensor = partial(torch.randn, device="cuda", dtype=torch.bfloat16)
B, H, seq_q, seq_k, D = 16, 16, 1024, 1024, 64

query = make_tensor((B, H, seq_q, D))
key = make_tensor((B, H, seq_k, D))
value = make_tensor((B, H, seq_k, D))

q_scale = torch.tensor(1, device="cuda", dtype=torch.float32)
k_scale = torch.tensor(1, device="cuda", dtype=torch.float32)
v_scale = torch.tensor(1, device="cuda", dtype=torch.float32)

q_fp8 = query.to(torch.float8_e4m3fn)
k_fp8 = key.to(torch.float8_e4m3fn)
v_fp8 = value.to(torch.float8_e4m3fn)
v_fp8 = v_fp8.to(torch.float8_e4m3fn).transpose(-1, -2).contiguous().transpose(-1, -2)

for _ in range(10):
    out = compiled_scale_flex(
        q_fp8,
        k_fp8,
        v_fp8,
        q_scale=q_scale,
        k_scale=k_scale,
        v_scale=v_scale,
        scale_func=per_tensor_scaling,
    )
    out_bf16 = compiled_flex(query, key, value)


time_fp8 = benchmark_cuda_function_in_microseconds(compiled_scale_flex, q_fp8, k_fp8, v_fp8, q_scale=q_scale, k_scale=k_scale, v_scale=v_scale, scale_func=per_tensor_scaling)
time_bf16 = benchmark_cuda_function_in_microseconds(compiled_flex, query, key, value)

print(f"fp8 time: {time_fp8} us")
print(f"bf16 time: {time_bf16} us")


```

Above is an example on how you can implement per-tensor scaling. That being said it is not as performant as FA3 and there are some more knobs that people would likely need/want and guarantees (e.g. that the out scaling is in fact fused) 

Will leave this open to track

CC @danielvegamyhre

**Comment by haochengxi** [comment]
I am curious about how FlexAttention quantizes the P matrix (after softmax)? If we want to accelerate the calculation of PV, both the P and V matrices need to be quantized into FP8 precision. But in the example above, it seems that P's quantization is not straightforward. Can you help explain more about this? Thanks!

**Comment by danielvegamyhre** [comment]
> I am curious about how FlexAttention quantizes the P matrix (after softmax)? If we want to accelerate the calculation of PV, both the P and V matrices need to be quantized into FP8 precision. But in the example above, it seems that P's quantization is not straightforward. Can you help explain more about this? Thanks!

We are currently thinking about this, it's tricky. Softmax is computed rowwise, so even if we use a rowwise scaling strategy, the values along the row won't be a valid probability distribution anymore, and it seems like this may adversely impact the numerics of the PV computation. 

Maybe with a granular, rowwise scaling strategy, the local statistics of each softmax row will be preserved enough for the numerics to be "good enough?"

I need to think about it some more, feel free to share your ideas/thoughts.
